libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+
+import * as crypto from 'crypto';
+import {
+  Async,
+  ColorValue,
+  ConsoleTerminalProvider,
+  InternalError,
+  ITerminal,
+  JsonObject,
+  Terminal
+} from '@rushstack/node-core-library';
+import { CollatedTerminal, CollatedWriter } from '@rushstack/stream-collator';
+import { DiscardStdoutTransform, PrintUtilities } from '@rushstack/terminal';
+import { SplitterTransform, TerminalWritable } from '@rushstack/terminal';
+
+import { CollatedTerminalProvider } from '../../utilities/CollatedTerminalProvider';
+import { ShellOperationRunner } from './ShellOperationRunner';
+import { OperationStatus } from './OperationStatus';
+import { CobuildLock, ICobuildCompletedState } from '../cobuild/CobuildLock';
+import { ProjectBuildCache } from '../buildCache/ProjectBuildCache';
+import { RushConstants } from '../RushConstants';
+import { IOperationSettings, RushProjectConfiguration } from '../../api/RushProjectConfiguration';
+import { getHashesForGlobsAsync } from '../buildCache/getHashesForGlobsAsync';
+import { ProjectLogWritable } from './ProjectLogWritable';
+import type { Operation } from './Operation';
+import type {
+  IOperationRunnerAfterExecuteContext,
+  IOperationRunnerBeforeExecuteContext
+} from './OperationRunnerHooks';
+import type { IOperationRunner, IOperationRunnerContext } from './IOperationRunner';
+import { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import type {
+  ICreateOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import { IRawRepoState, ProjectChangeAnalyzer } from '../ProjectChangeAnalyzer';
+import type { OperationMetadataManager } from './OperationMetadataManager';
+import type { BuildCacheConfiguration } from '../../api/BuildCacheConfiguration';
+import { CobuildConfiguration } from '../../api/CobuildConfiguration';
+import { DisjointSet } from '../cobuild/DisjointSet';
+
+const PLUGIN_NAME: 'CacheablePhasedOperationPlugin' = 'CacheablePhasedOperationPlugin';
+
+export interface IOperationBuildCacheContext {
+  isCacheWriteAllowed: boolean;
+  isCacheReadAllowed: boolean;
+  isSkipAllowed: boolean;
+  projectBuildCache: ProjectBuildCache | undefined;
+  cobuildLock: CobuildLock | undefined;
+  // The id of the cluster contains the operation, used when acquiring cobuild lock
+  cobuildClusterId: string | undefined;
+  // Controls the log for the cache subsystem
+  buildCacheTerminal: ITerminal | undefined;
+  buildCacheProjectLogWritable: ProjectLogWritable | undefined;
+}
+
+export class CacheableOperationPlugin implements IPhasedCommandPlugin {
+  private _buildCacheContextByOperationRunner: Map<IOperationRunner, IOperationBuildCacheContext> = new Map<
+    IOperationRunner,
+    IOperationBuildCacheContext
+  >();
+
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tapPromise(
+      PLUGIN_NAME,
```

Comment:
This should probably tap `beforeExecuteOperations` instead of `createOperations`.

This is particularly important because other plugins in the ecosystem might have replaced `ShellOperationRunner` instances with other runners in between `createOperations` and `beforeExecuteOperations`, depending on how the sequencing was implemented.

Reply:
I've changed the logic to tap `beforeExecuteOperations`. But the current logic only apply related logic to `ShellOperationRunner` via `_applyShellOperationRunner`. Maybe it should be refactored to apply to a more generic "OperationRunner" (?)